common/src/main/java/io/netty/util/ByteString.java

```diff
@@ -0,0 +1,653 @@
+        if (copy) {
+            this.value = checkNotNull(value, "value").clone();
+        } else {
+            this.value = checkNotNull(value, "value");
+        }
+    }
+
+    /**
+     * Initialize this byte string based upon a range of a byte array. A copy will be made.
+     */
+    public ByteString(byte[] value, int start, int length) {
+        this(value, start, length, true);
+    }
+
+    /**
+     * Construct a new {@link BinaryString} object from a {@code byte[]} array.
+     * @param copy {@code true} then a copy of the memory will be made. {@code false} the underlying memory
+     * will be shared. If this shared memory changes then {@link #arrayChanged()} must be called.
+     */
+    public ByteString(byte[] value, int start, int length, boolean copy) {
+        if (start < 0 || start > checkNotNull(value, "value").length - length) {
+            throw new IndexOutOfBoundsException("expected: " + "0 <= start(" + start + ") <= start + length(" + length
+                            + ") <= " + "value.length(" + value.length + ')');
+        }
+
+        if (copy || start != 0 || length != value.length) {
+            this.value = Arrays.copyOfRange(value, start, start + length);
+        } else {
+            this.value = value;
+        }
+    }
+
+    /**
+     * @see {@link #getBytes(ByteBuffer)}
+     */
+    public ByteString(ByteBuffer value) {
+        this.value = getBytes(value);
+    }
+
+    /**
+     * @see {@link #getBytes(ByteBuffer, int, int)}
+     */
+    public ByteString(ByteBuffer value, int start, int length) {
+        this.value = getBytes(value, start, length, true);
+    }
+
+    /**
+     * @see {@link #getBytes(ByteBuffer, int, int, boolean)}
+     */
+    public ByteString(ByteBuffer value, int start, int length, boolean copy) {
+        this.value = getBytes(value, start, length, copy);
+    }
+
+    /**
+     * @see {@link #getBytes(char[], Charset)}
+     */
+    public ByteString(char[] value, Charset charset) {
+        this.value = getBytes(value, charset);
+    }
+
+    /**
+     * @see {@link #getBytes(char[], Charset, int, int)}
+     */
+    public ByteString(char[] value, Charset charset, int start, int length) {
+        this.value = getBytes(value, charset, start, length);
+    }
+
+    /**
+     * @see {@link #getBytes(CharSequence, Charset)}
+     */
+    public ByteString(CharSequence value, Charset charset) {
+        this.value = getBytes(value, charset);
+    }
+
+    /**
+     * @see {@link #getBytes(CharSequence, Charset, int, int)}
+     */
+    public ByteString(CharSequence value, Charset charset, int start, int length) {
+        this.value = getBytes(value, charset, start, length);
+    }
+
+    /**
+     * Create a copy of the underlying storage from {@link value} into a byte array.
+     * The copy will start at {@link ByteBuffer#position()} and copy {@link ByteBuffer#remaining()} bytes.
+     */
+    public static byte[] getBytes(ByteBuffer value) {
```

Comment:
Let's make them private and consider moving them somewhere else and making them public in a later discussion.